hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/CryptoGetAccountInfoHandler.java

```diff
@@ -58,13 +108,258 @@ public Response createEmptyResponse(@NonNull final ResponseHeader header) {
+            @NonNull final TokensConfig tokenConfig,
+            @NonNull final Account account,
+            @NonNull final ReadableTokenStore readableTokenStore,
+            @NonNull final ReadableTokenRelationStore tokenRelationStore) {
+        requireNonNull(account);
+        requireNonNull(tokenConfig);
+        requireNonNull(readableTokenStore);
+        requireNonNull(tokenRelationStore);
+        final var ret = new ArrayList<TokenRelationship>();
+        var tokenNum = account.headTokenNumber();
+        int count = 0;
+        Optional<TokenRelation> tokenRelation;
+        Token token; // token from readableToken store by tokenID
+        TokenID tokenID; // build from tokenNum
+        AccountID accountID; // build from accountNumber
+        TokenRelationship tokenRelationship; // created TokenRelationship object
+        while (tokenNum != 0 && count < tokenConfig.maxRelsPerInfoQuery()) {
+            accountID =
+                    AccountID.newBuilder().accountNum(account.accountNumber()).build();
+            tokenID = TokenID.newBuilder().tokenNum(tokenNum).build();
+            tokenRelation = tokenRelationStore.get(accountID, tokenID);
+            if (tokenRelation.isPresent()) {
+                token = readableTokenStore.get(tokenID);
+                if (token != null) {
+                    tokenRelationship = TokenRelationship.newBuilder()
+                            .tokenId(TokenID.newBuilder().tokenNum(tokenNum).build())
+                            .symbol(token.symbol())
+                            .balance(tokenRelation.get().balance())
+                            .decimals(token.decimals())
+                            .kycStatus(tokenRelation.get().kycGranted() ? GRANTED : KYC_NOT_APPLICABLE)
+                            .freezeStatus(tokenRelation.get().frozen() ? FROZEN : FREEZE_NOT_APPLICABLE)
+                            .automaticAssociation(tokenRelation.get().automaticAssociation())
+                            .build();
+                    ret.add(tokenRelationship);
+                }
+                tokenNum = tokenRelation.get().nextToken();
+            } else {
+                break;
+            }
+            count++;
+        }
+        return ret;
+    }
+
+    private String getContractAccountId(Account account, final Bytes alias) {
+        if (alias.toByteArray().length == EVM_ADDRESS_SIZE) {
+            return CommonUtils.hex(alias.toByteArray());
+        }
+        // If we can recover an Ethereum EOA address from the account key, we should return that
+        final var evmAddress = tryAddressRecovery(account.key(), EthSigsUtils::recoverAddressFromPubKey);
+        if (evmAddress != null && evmAddress.length == EVM_ADDRESS_LEN) {
+            return Bytes.wrap(evmAddress).toHex();
+        } else {
+            return CommonUtils.hex(asEvmAddress(account.accountNumber()));
+        }
+    }
+
+    private static byte[] asEvmAddress(final long num) {
```

Comment:
This would be a good candidate function to add to CommonUtils or similar shared location

Reply:
moved it to CommonUtils